lui.el

```diff
@@ -1333,5 +1390,135 @@ which see for an explanation of the argument BUFFER-STRING."
                            front-sticky t))))
 
 
+;;;;;;;;;;;;;;;;;;
+;;; Navigation ;;;
+;;;;;;;;;;;;;;;;;;
+
+(defun lui-at-message-p ()
+  "Check if point is on a message."
+  (get-text-property (point) 'lui-message-id))
+
+(defun lui-beginning-of-message-p ()
+  "Check if point is at the beginning of a message."
+  (not (equal (get-text-property (point) 'lui-message-id)
+              (get-text-property (1- (point)) 'lui-message-id))))
+
+(defun lui-beginning-of-message ()
+  "Move point to the beginning of the message at point."
+  (goto-char (previous-single-property-change (point) 'lui-message-id)))
+
+(defun lui-forward-message ()
+  "Move point to the next message in the buffer and return point.
+If there is no next message, move to the end of the buffer instead."
+  (let ((current-id (get-text-property (point) 'lui-message-id))
+        (next-point
+         (next-single-property-change (point) 'lui-message-id)))
+    (if next-point
+        (let ((message-id
+               (get-text-property next-point 'lui-message-id)))
+          (goto-char next-point)
+          (when (or (not (or current-id message-id))
+                    (and current-id (not message-id))
+                    (and current-id message-id
+                         (= current-id message-id)))
+            (lui-forward-message)))
+      (goto-char (point-max)))
+    (point)))
+
+(defun lui-backward-message ()
+  "Move point to the previous message in the buffer and return point.
+If there is no previous message, move to the beginning of the
+buffer instead."
+  (let ((current-id (get-text-property (point) 'lui-message-id))
+        (prev-point
+         (previous-single-property-change (point) 'lui-message-id)))
+    (if prev-point
+        (let ((message-id
+               (get-text-property prev-point 'lui-message-id)))
+          (goto-char prev-point)
+          (when (or (not (or current-id message-id))
+                    (and current-id (not message-id))
+                    (and current-id message-id
+                         (= current-id message-id)))
+            (lui-backward-message)))
+      (goto-char (point-min)))
+    (point)))
+
+
+;;;;;;;;;;;;;;;
+;;; Editing ;;;
+;;;;;;;;;;;;;;;
+
+(defun lui-recover-output-marker ()
+  "Reset the output marker to just before the lui prompt."
+  (let ((input-position (marker-position lui-input-marker)))
+    (set-marker lui-output-marker
+                (field-beginning (1- input-position)))))
+
+(defun lui-build-plist (keys)
+  "Build a plist with KEYS taken from current text properties."
+  (let (result)
+    (mapc (lambda (key)
+            (let ((value (get-text-property (point) key)))
+              (if value
+                  (setq result (plist-put result key value)))))
+          keys)
+    result))
```

Comment:
oh yes, I agree. I almost never use `dolist`, somehow I learned that was a bad idea. I should change that.